stock_card_report/reports/stock_card_report.py

```diff
@@ -0,0 +1,107 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class StockCardView(models.TransientModel):
+    _name = 'stock.card.view'
+    _inherit = 'stock.move'
+    _order = 'id'
+
+    is_initial = fields.Boolean()
+    product_in = fields.Float()
+    product_out = fields.Float()
+
+
+class StockCardReport(models.TransientModel):
+    _name = 'report.stock.card.report'
+
+    # Filters fields, used for data computation
+    date_from = fields.Date()
+    date_to = fields.Date()
+    product_ids = fields.Many2many(
+        comodel_name='product.product',
+    )
+    location_id = fields.Many2one(
+        comodel_name='stock.location',
+    )
+
+    # Data fields, used to browse report data
+    results = fields.Many2many(
+        comodel_name='stock.card.view',
+        compute='_compute_results',
+        help='Use compute fields, so there is nothing store in database',
+    )
+
+    @api.multi
+    def _compute_results(self):
+        self.ensure_one()
+        change = False
+        if not self.date_from:
+            change = True
+            self.date_from = '0001-01-01'
+        if not self.date_to:
+            self.date_to = fields.Date.context_today(self)
+        self._cr.execute("""
+            SELECT move.id, move.name, move.date, move.product_id,
+                move.product_qty, move.product_uom_qty, move.product_uom,
+                move.location_id, move.location_dest_id, move.reference,
+                case when move.location_dest_id = %s
+                    then move.product_qty end as product_in,
+                case when move.location_id = %s
+                    then move.product_qty end as product_out,
+                case when move.date < %s then True else False end as is_initial
+            FROM stock_move move
+            WHERE (move.location_id = %s or move.location_dest_id = %s)
+                and move.state = 'done' and move.product_id in %s
+                and CAST(move.date AS date) <= %s
+            ORDER BY move.date
+        """, (
+            self.location_id.id, self.location_id.id, self.date_from,
+            self.location_id.id, self.location_id.id,
+            tuple(self.product_ids.ids+[0]), self.date_to))
+        stock_card_results = self._cr.dictfetchall()
+        ReportLine = self.env['stock.card.view']
+        for line in stock_card_results:
+            self.results += ReportLine.new(line)
+        if change:
+            self.date_from = not change
+
+    @api.multi
+    def _get_initial(self, product_line):
+        product_input_qty = sum(product_line.mapped('product_in'))
+        product_output_qty = sum(product_line.mapped('product_out'))
+        return product_input_qty - product_output_qty
+
+
+class StockCardReportCompute(models.TransientModel):
+    _inherit = 'report.stock.card.report'
+
+    @api.multi
+    def print_report(self, report_type='qweb'):
+        self.ensure_one()
+        if report_type == 'xlsx':
```

Comment:
To get the action, can we short cut to the ext_id,
```suggestion
action = report_type == 'xlsx' and self.env.ref('stock_card_report.action_stock_card_report_xlsx') or \
  self.env.ref('stock_card_report.action_stock_card_report_pdf')
```
```